File: app/components/folio/console/aasm/email_modal_component.rb
# frozen_string_literal: true

# Opened by Folio::Console::StateCell's trigger for an aasm event declared with
# `email_modal: true`. The console user confirms the state change and, if they
# want, sends the record's e-mail address a message about it. The form posts
# straight to Folio::Console::Api::AasmController#event, which performs the
# transition and hands back the re-rendered state cell.
class Folio::Console::Aasm::EmailModalComponent < Folio::Console::ApplicationComponent
  include SimpleForm::ActionViewExtensions::FormHelper

  CLASS_NAME = "f-c-aasm-email-modal"

  def initialize
  end

  private
    def data
      stimulus_controller(CLASS_NAME,
                          action: {
                            "folioConsoleAasmEmailModalOpen" => "openFromEvent",
                          },
                          values: {
                            title: t(".title"),
                            send_email_label: t(".send_email_label"),
                          },
                          classes: %w[loading])
    end

    def form(&block)
      simple_form_for("", url: controller.folio.event_console_api_aasm_path, html: form_html, &block)
    end

    def form_html
      {
        class: "#{CLASS_NAME}__form",
        data: stimulus_action(submit: "onFormSubmit",
                              change: "onFormChange",
                              input: "onFormChange"),
      }
    end

    def key_hidden_field(f, key)
      f.hidden_field key,
                     value: nil,
                     class: "#{CLASS_NAME}__hidden",
                     data: stimulus_target("hidden")
    end

    def checkbox_input_html
      {
        class: "#{CLASS_NAME}__checkbox",
        checked: true,
        data: stimulus_target("checkbox"),
      }
    end

    def checkbox_label_html
      { data: stimulus_target("checkboxLabel") }
    end

    def subject_input_html
      {
        class: "#{CLASS_NAME}__subject",
        placeholder: t(".placeholders.subject"),
        data: stimulus_target("subject"),
      }
    end

    def text_input_html
      {
        class: "#{CLASS_NAME}__text",
        placeholder: t(".placeholders.text"),
        rows: 8,
        data: stimulus_target("text"),
      }
    end
end

File: app/components/folio/console/aasm/email_modal_component.js
window.FolioConsole = window.FolioConsole || {}
window.FolioConsole.AasmEmailModal = window.FolioConsole.AasmEmailModal || {}

// Called by the f-c-state controller when the clicked event declares
// `email_modal: true`. detail carries the trigger's data attributes.
window.FolioConsole.AasmEmailModal.open = (detail) => {
  const modal = document.querySelector('.f-c-aasm-email-modal')

  if (!modal) {
    throw new Error('Missing .f-c-aasm-email-modal. Render Folio::Console::Aasm::EmailModalComponent in the console layout.')
  }

  modal.dispatchEvent(new window.CustomEvent('folioConsoleAasmEmailModalOpen', { bubbles: true, detail }))
}

window.Folio.Stimulus.register('f-c-aasm-email-modal', class extends window.Stimulus.Controller {
  static values = {
    title: String,
    sendEmailLabel: String
  }

  static targets = [
    'title',
    'checkbox',
    'checkboxLabel',
    'subject',
    'text',
    'submit',
    'hidden'
  ]

  static classes = ['loading']

  disconnect () {
    this.trigger = null
  }

  openFromEvent (e) {
    const { trigger, targetStateName, email, klass, id, aasmEvent, emailSubject, emailText } = e.detail

    this.trigger = trigger

    this.titleTarget.innerText = this.titleValue.replace('{STATE_NAME}', targetStateName)
    this.checkboxTarget.checked = true
    this.checkboxLabelTarget.innerText = this.sendEmailLabelValue.replace('{EMAIL}', email || '')

    this.setHidden('klass', klass)
    this.setHidden('aasm_event', aasmEvent)
    this.setHidden('id', id)
    this.setHidden('email', email)

    this.subjectTarget.value = emailSubject || ''
    this.textTarget.value = emailText || ''

    this.validate()

    window.Folio.Modal.open(this.element)

    if (this.subjectTarget.value) {
      this.textTarget.focus()
    } else {
      this.subjectTarget.focus()
    }
  }

  setHidden (key, value) {
    const input = this.hiddenTargets.find((el) => el.name === key)
    if (input) input.value = value == null ? '' : value
  }

  // Sending an empty message is never what the user meant; not sending one at
  // all is fine. When the checkbox is on, both subject and body must be present
  // — the endpoint rejects the request otherwise.
  validate () {
    const emailEnabled = this.checkboxTarget.checked
    const missingSubject = !this.subjectTarget.value.trim()
    const missingBody = !this.textTarget.value.trim()

    this.submitTarget.disabled = emailEnabled && (missingSubject || missingBody)
  }

  onFormChange () {
    this.validate()
  }

  onFormSubmit (e) {
    e.preventDefault()

    if (this.element.classList.contains(this.loadingClass)) return

    const form = e.target
    const trigger = this.trigger
    const payload = window.Folio.formToHash(form)

    this.element.classList.add(this.loadingClass)
    // Close before apiPost so Folio.Api's auto-flash lands in the layout,
    // not inside this still-open modal.
    window.Folio.Modal.close(this.element)

    window.Folio.Api.apiPost(form.action, payload).then((res) => {
      if (trigger) {
        const state = trigger.closest('.f-c-state')
        if (state && res && res.data) state.outerHTML = res.data
      }

      this.trigger = null
      form.reset()
    }).catch((error) => {
      window.Folio.Modal.open(this.element)
      document.dispatchEvent(new window.CustomEvent('folio:flash', {
        bubbles: true,
        detail: { flash: { content: error.message, variant: 'danger' } }
      }))
    }).finally(() => {
      this.element.classList.remove(this.loadingClass)
    })
  }
})
